### PR TITLE
ARROW-14431: [C++][Gandiva] Implement AES ENCRYPT and AES DECRYPT functions 

### PR DESCRIPTION
**Pull request by augustoasilva**
Encrypt, or decrypt, input using AES.

### CODE REVIEW

File: cpp/src/gandiva/encrypt_utils.cc [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+#include "encrypt_utils.h"
+
+#include "gandiva/gdv_function_stubs.h"
+
+namespace gandiva {
+
+int32_t aes_encrypt(int64_t context, unsigned char* plaintext, int32_t plaintext_len,
```
Comment (vvellanki) [inline cpp/src/gandiva/encrypt_utils.cc:24]: This is going to perform poorly. This function is called for each and every row in the query.

This has too much overhead - please talk to Projjal about how to implement this correctly

Comment (vvellanki) [reply cpp/src/gandiva/encrypt_utils.cc:24]: This is ok for now

File: cpp/src/gandiva/function_registry_encrypt.cc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#include "function_registry_encrypt.h"
+#include "gandiva/function_registry_common.h"
+
+namespace gandiva {
+
+std::vector<NativeFunction> GetEncryptFunctionRegistry() {
+  static std::vector<NativeFunction> encrypt_fn_registry_ = {
+      NativeFunction("aes_encrypt", {}, DataTypeVector{utf8(), utf8()}, utf8(),
+                     kResultNullIfNull, "gdv_fn_aes_encrypt",
+                     NativeFunction::kNeedsContext),
```
Comment (vvellanki) [inline cpp/src/gandiva/function_registry_encrypt.cc:27]: The implementation of these functions does not use context. Why is this function marked as kNeedsContext?

File: cpp/src/gandiva/encrypt_utils_test.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+#include <unordered_set>
+
+#include "gandiva/execution_context.h"
+
+TEST(TestShaEncryptUtils, TestAesEncryptDecrypt) {
+  gandiva::ExecutionContext ctx;
+  auto ctx_ptr = reinterpret_cast<int64_t>(&ctx);
+
+  auto* key = (unsigned char*)"12345abcde";
```
Comment (vvellanki) [inline cpp/src/gandiva/encrypt_utils_test.cc:30]: Can you test with large key lengths? Please add test cases with 16, 32, 64, 128, 192, and 256 byte keys

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 4]
```diff
@@ -794,6 +796,60 @@ const char* gdv_fn_initcap_utf8(int64_t context, const char* data, int32_t data_
...
+const char* gdv_fn_aes_encrypt(int64_t context, const char* data, int32_t data_len,
+                               const char* key_data, int32_t key_data_len,
+                               int32_t* out_len) {
+  if (data_len < 0) {
+    gdv_fn_context_set_error_msg(context, "Invalid data length to be encrypted");
+    *out_len = 0;
+    return "";
+  }
+
+  unsigned char encrypted[256];
```
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:810]: Why do we need this temporary buffer? Why not allocate memory first and pass a pointer to that buffer to aes_encrypt?

File: cpp/src/gandiva/function_registry_encrypt.cc (hunk 2 above)
Comment (augustoasilva) [reply cpp/src/gandiva/function_registry_encrypt.cc]: Now it uses context and has a better error handling, so I also added the flag kCanReturnErrors

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 4 above)
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc]: We dont need it anymore

File: cpp/src/gandiva/encrypt_utils_test.cc (hunk 3 above)
Comment (augustoasilva) [reply cpp/src/gandiva/encrypt_utils_test.cc]: Tested with 8, 16, 32, 64, 128, 192 and 256 byte keys.

**Comment by vvellanki** [comment]
Looks good to me. Lets get the build issues fixed and merge the change

**Comment by vvellanki** [comment]
The build is still failing

**Review by vvellanki — APPROVED** [review]

**Comment by kszucs** [comment]
@pravindra @vvellanki it broke the [docs build](https://github.com/apache/arrow/runs/4456015064?check_suite_focus=true) where `CMAKE_UNITY_BUILD` is enabled.

**Comment by lidavidm** [comment]
Filed ARROW-15028 for this

**Pull request closed without merging**